---
title: "Analysis of data obtained from SRA studies"
output:
  pdf_document: default
  html_notebook: default
  html_document: default
---

```{r, echo = F}
knitr::opts_chunk$set(fig.width=12, fig.height=9, fig.keep='all', fig.show='asis', tidy = TRUE, fig.pos='htb!')
options(width = 60)
```

In the first part of my PhD project, I have obtained transcriptomic data from malaria studies from SRA database to reconstruct an impartial overview of the core interaction modules between host and *Plasmodium sp.*. In this document, I have analysed RNA-seq data these studies after mapping the data onto corresponding reference genomes. In total, I have 30 studies, 10 each for mouse (*Mus musculus*), human and monkey (*Macaca mulatta*). 

```{r, echo=FALSE, eval=TRUE}
# Make colour palette for hp pairs

# install.packages("randomcoloR")
# library(randomcoloR)
# n <- 8
# palette <- distinctColorPalette(n)
# 
# hp_pairs <- c("humanPvivax", "humanPfalciparum", "humanPberghei", "mousePyoelii", "mousePchabaudi", "mousePberghei", "monkeyPcynomolgi", "monkeyPcoatnei")
# 
# hp_pair_colour <- data.frame(hp_pairs, palette)
# hp_pair_colour <- data.frame(hp_pair_colour, Shape=c(0,1,2,3,4,5,6,20))

# run with knitr::knit("filename.RData")
# rmarkdown::render()
```

Plot 1a. Host expression vs parasite expression

This plot has host expression % on one axis and parasite expression % on the other of every run in all the studies. The points should form a diagonal line with negative slope (-1), because the sum of the values on x and y axis should be 100. The number of points should be more close to host 100% expression and less near parasite 100% expression. Near the centre there should be less points, as not a lot of runs had 50-50 host-parasite expression. The points at the centre have 50-50 expression, hence the "best".

```{r, echo=FALSE, eval=FALSE}
# First get all the study IDs and the associated hosts and parasites. This info is stored in a comma separated text file named positive_experiments.txt. From this file we get name of host, of parasite and study ID.

library(dplyr)
library(tibble)

positive_experiments <- read.table("/SAN/Plasmo_compare/SRAdb/Input/positive_experiments.txt", header = F, sep = '\t')
studyIDs <- as.character(positive_experiments[,1])
#studyIDs <- c("ERP020067", "SRP018945", "SRP066796", "ERP106451")

# Second, we want the run IDs for each of the studies we read from the above file. So we go into the folder of each study and read the run IDs from runs_<study>.txt

getAllHPexpression <- function(study)
{ 
  print(study)
  study_df <- data.frame()
  
  # the following statement gives us both run ID and the study ID
  runIDs <- read.table(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/runs_",study,".txt",collapse=''), header = F, sep = ',')
  
  # number of runs
  number_of_runs <- nrow(runIDs)
  
  # get hp_percent_<study>.txt
  hp_percent_study <- read.table(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/hp_percent_",study,".txt",collapse=''), header = T, sep = '\t')
  
            for(i in 1:number_of_runs)
    {
      # get runID
      runID <- as.character(runIDs[i,1])
      
      # get the host and parasite expression percentages for the runID from hp_percent_study
      # host_percent <- hp_percent_study[grep(runID, hp_percent_study[,1]), 2]
      # parasite_percent <- hp_percent_study[grep(runID, hp_percent_study[,1]), 3]
      # 
      # host <- positive_experiments[grep(study,positive_experiments[,1]),2]
      # parasite <- positive_experiments[grep(study,positive_experiments[,1]),3]
      # 
      # count_file <- read.table(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/countWithGFF3_",runID,".txt",collapse=''), header = T, sep = '\t')
      # if(file.exists(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study[i],"/",runID,"_",study[i],".final.out",collapse='')))
      # {
      #   inputReadLine <- readLines(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/",runID,"_",study,".final.out",collapse=''), n=6)[6]
      #   mapPercentLine <- readLines(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/",runID,"_",study,".final.out",collapse=''), n=10)[10]
      #   
      #   multipleReadsLine <- readLines(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/",runID,"_",study,".final.out",collapse=''), n=26)[c(24,26)]
      #   
      #   unmappedReadsLine <- readLines(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/",runID,"_",study,".final.out",collapse=''), n=31)[c(29,30,31)]
      # 
      #   inputReads <- as.numeric(strsplit(inputReadLine, split="\t")[[1]][2])
      #   map_percent <- as.numeric(strsplit(strsplit(mapPercentLine, split="%")[[1]][2], split="\t")[[1]][2])
      #   multipleReads <- sum(sapply(multipleReadsLine, function(x) as.numeric(strsplit(x, split="\t")[[1]][2])))
      #   unmappedReads <- sum(sapply(unmappedReadsLine, function(x) (as.numeric(strsplit(strsplit(x, split="%")[[1]][2], split="\t")[[1]][2])/100) * inputReads))
      # }else
      # {
      #   inputReads <- "NA"
      #   map_percent <- "NA"
      #   multipleReads <- "NA"
      #   unmappedReads <- "NA"
      # }
      # 
      # parasite_rows <- as.numeric(grep("P+", count_file[,1]))
      # host_count <- count_file[-c(parasite_rows),]
      # parasite_count <- count_file[parasite_rows,]
      #   
      # host_sum <- sum(host_count[,6])
      # parasite_sum <- sum(parasite_count[,6])
      # 
      protein_coding <- read.delim(paste0("/SAN/Plasmo_compare/SRAdb/Output/", study, "/", study, "_coding_genes.txt", collapse = '')) %>%
        # take the current run
        select(contains(runID)) %>% 
        tibble::rownames_to_column("Gene")
      
      if(ncol(protein_coding) == 2)
      {
        parasite_rows_prot_cod <- as.numeric(grep("P+", protein_coding[,"Gene"]))
        host_prot_cod <- protein_coding[-c(parasite_rows_prot_cod),]
        parasite_prot_cod <- protein_coding[parasite_rows_prot_cod,]
        
        # host_count_prot_cod <- sum(host_prot_cod[,2]) #(total number of reads)
        # parasite_count_prot_cod <- sum(parasite_prot_cod[,2])
        # 
        host_gene_prot_cod <- length(which(host_prot_cod[,2]>0)) # Total number of gene with count > 0
        parasite_gene_prot_cod <- length(which(parasite_prot_cod[,2]>0))
        
        # actual_host <- hp_percent_study[grep(runID, hp_percent_study[,1]),5]
        # actual_para <- hp_percent_study[grep(runID, hp_percent_study[,1]),6]
        # 
        study_df[i,1] <- runID
        study_df[i,2] <- study
        # study_df[i,3] <- host
        # study_df[i,4] <- host_percent
        # study_df[i,5] <- parasite
        # study_df[i,6] <- parasite_percent
        # study_df[i,7] <- sum(count_file[,6]) # sum(counts)
        # study_df[i,8] <- inputReads # number of input reads
        # study_df[i,9] <- map_percent # map_percent
        # study_df[i,10] <- paste0(host,parasite, collapse='')
        # study_df[i,11] <- host_sum
        # study_df[i,12] <- parasite_sum
        # study_df[i,13] <- multipleReads
        # study_df[i,14] <- unmappedReads
        # study_df[i,15] <- actual_para # actual proportion of parasites, taking into account unique map%
        # study_df[i,16] <- actual_host
        # study_df[i,17] <- host_count_prot_cod
        # study_df[i,18] <- parasite_count_prot_cod
        study_df[i,3] <- host_gene_prot_cod
        study_df[i,4] <- parasite_gene_prot_cod
      }
    }
  return (study_df)
}

allHPexpression <- data.frame()
for(j in 1:length(studyIDs))
{
  allHPexpression <- rbind(allHPexpression, getAllHPexpression(studyIDs[j]))
}
colnames(allHPexpression) <- c("RunID", "Study", "NumberProtCodGenesHost", "NumberProtCodGenesPara") #  "Host", "Host_percent", "Parasite", "Parasite_percent", "CountSum", "InputReads",  "MapPercent" ,"HostParasite", "HostSum", "ParasiteSum", "MultipleMapReads", "UnmappedReads", "ActualParasite", "ActualHost","ProteinCodHost", "ProteinCodPara"
```

```{r, echo = F}

# Plot

require(ggplot2)
#load("/SAN/Plasmo_compare/SRAdb/allHPexpression.RData")

allHPexpression <- read.delim("/SAN/Plasmo_compare/SRAdb/Output/allHPexp.txt")

# col=c("mousePyoelii"="red", "monkeyPcynomolgi"="blue", "humanPfalciparum"="green")
# shape=c("mousePyoelii"=0, "monkeyPcynomolgi"=1)

allHPexpression_ggplot <- ggplot(allHPexpression, aes(Parasite_percent, Host_percent, col=HostParasite, shape=HostParasite))

allHPexpression_gg <- allHPexpression_ggplot + geom_point(alpha = 0.5) + geom_hline(yintercept=50, color="brown") + geom_vline(xintercept=50, color="brown") + theme_bw() + xlim(0,100) + ylim(0,100) + ggtitle("Host expression vs parasite expression") + xlab("Parasite gene expression percentage") + ylab("Host gene expression percentage") + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18)) + labs(color='Host-parasite\npairs') + labs(shape='Host-parasite\npairs') + scale_shape_manual(values = 0:7)
allHPexpression_gg
```

1b. The plot below shows percentage host vs percentage parasite that was uniquely mapped in a run in a study. This refers to (host_percent_overall * uniquely_mappedReads_percent).

```{r, echo = F}

# Plot

require(ggplot2)

# col=c("mousePyoelii"="red", "monkeyPcynomolgi"="blue", "humanPfalciparum"="green")
# shape=c("mousePyoelii"=0, "monkeyPcynomolgi"=1)

allHPexpression_ggplot <- ggplot(allHPexpression, aes(ActualParasite, ActualHost, col=HostParasite, shape=HostParasite))

allHPexpression_gg <- allHPexpression_ggplot + geom_point(alpha = 0.5) + geom_hline(yintercept=50, color="brown") + geom_vline(xintercept=50, color="brown") + theme_bw() + xlim(0,100) + ylim(0,100) + ggtitle("Host expression vs parasite expression (uniquely mapped)") + xlab("Parasite gene expression percentage") + ylab("Host gene expression percentage") + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18)) + labs(color='Host-parasite\npairs') + labs(shape='Host-parasite\npairs') + scale_shape_manual(values = 0:7)
allHPexpression_gg
```

Here, the diagonal is below the 100-100 line because the no run had 100% unique mapping of reads. However, one has to verify why some runs are quite far away from the diagonal cloud. There seems to be a big cloud of a mousePyoelii study, which could be SRP075802 because it was a microbiome study. Other major cloud is from humanPfalciparum (mostly from DRP000987), a few from humanPberghei (SRP034011) and a dispersed cloud from mousePchabaudi (ERP004042, ERP005730). A total of around 120 runs have unique map percent < 60%, many of which are humanPfalciparum, fewer mousePchabaudi and fewer still from humanPberghei and monkeyPcynomolgi (SRP106638). The two human studies that involved dual RNA-seq have less uniquely mapped proportions. Same for ERP004042. [Yet to find out reason. mousePchabaudi papers are not available yet]

Plot 2. Parasite distribution in every study.

This is a violin plot to show the distribution of parasite proportion in every study. Here, on the x-axis is parasite proportion ranging from 0 to 100% and the studies are named on the y-axis. 

```{r, echo=F, eval=TRUE}

# For this we require two columns of allHPexpression - Study and Parasite_percent

parasiteProportion_ggplot <- ggplot(allHPexpression, aes(x = as.factor(Study), y = Parasite_percent, col=HostParasite, fill=HostParasite, shape=HostParasite))

parasiteProportion_gg <- parasiteProportion_ggplot + geom_violin(scale="width", alpha=0.3) + ylim(0,100) + geom_jitter(height = 0, width = 0.1, size = 2.5) + ggtitle("Parasite distribution in studies") + xlab("SRA study ID") + ylab("Parasite gene expression percentage") + theme(axis.text=element_text(size=14), axis.title=element_text(size=16), plot.title = element_text(size=20), axis.text.x = element_text(angle = 90, vjust = 1, size = 7)) + theme_classic() + labs(color='Host-parasite\npairs') + labs(shape='Host-parasite\npairs') + labs(fill='Host-parasite\npairs') + theme(strip.background = element_rect(fill=c("aliceblue"), colour = "lightblue"), strip.text.x = element_text(size = 12, colour = "black")) + facet_wrap(~Host, scales = "free") + coord_flip() + scale_shape_manual(values = 0:7)

parasiteProportion_gg
```

It is evident here that human studies have provided the least cross-species transcriptomic data, followed by monkey studies and then mouse studies, being the best source of cross-species transcriptomic data.

Possible reasons for the variations:

1. Humans show more immunity, because the others, unlike the humans, were lab animals.
2. All human studies were from endemic regions: Columbia, Indonesia, Gambia, Mali, Tanzania.
3. "Overwhelming abundance of human transcripts in whole blood un-enriched for parasite material, with adequate library sizes for P. falciparum only observed at high parasite densities (>10,000 parasites/microlitre)", obtained from SRP029990 [doi:10.1038/srep31291]

\newpage

Plot 3. Distribution of parasite proportions across all runs in all studies.

This is a scatter plot showing how the total number of overlapped exonic regions (result from countOverlaps()) relate with parasite proportions for a run in a given host-parasite pair.

```{r, echo = F, eval = TRUE}
parasiteProportionCluster_ggplot <- ggplot(allHPexpression, aes(x=Parasite_percent, y=log10(CountSum), col=HostParasite, shape=HostParasite))

parasiteProportionCluster_gg <- parasiteProportionCluster_ggplot + geom_point() + xlim(0,100) + labs(color='Host-parasite\npairs') + labs(shape='Host-parasite\npairs') + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18)) + theme_bw() + scale_shape_manual(values = 0:7)

parasiteProportionCluster_gg
```

<!-- Plot 4. Pyramid plot for overlapping exonic regions on host and parasite genomes -->

<!-- In this plot, I want to see the number of exonic regions that overlapped with the host and parasite genomes in a study. We can make the sum of counts for all host genes and all parasite genes and make a pyramid plot for these values. (Simply doing host_percent*CountSum is sometimes not accurate.) -->

<!-- Is it possible to find out the actual number of input reads that mapped onto each of the genomes? -->

```{r, echo = F, eval = TRUE}
# 
# HostParasiteCountPyramid_ggplot <- function(study)
# {
#   require(reshape2)
#   for(i in 1:length(study))
#   {
#     study <- study[i]
#   # get rows for the study
#    studyHPexpression <- allHPexpression[grep(study, allHPexpression[,2]),c("RunID", "Study", "Host", "Parasite", "HostParasite", "CountSum", "InputReads", "MapPercent", "Host_percent", "Parasite_percent", "HostSum", "ParasiteSum")]
#   studyHPexp_melt <- melt(studyHPexpression, id=c("RunID", "Study", "Host", "Parasite", "HostParasite", "CountSum", "InputReads", "MapPercent", "Host_percent", "Parasite_percent"))
#   
#   
#   plot_gg <- ggplot(data = studyHPexp_melt, mapping = aes(x = RunID, fill = variable, y = ifelse(test = variable == "HostSum", yes = -value, no = value))) + geom_bar(stat = "identity", width=0.5) + scale_y_continuous(labels = abs, limits = max(studyHPexpression[,"HostSum"]) * c(-1,1)) + labs(y = paste0("Overlaps in ", study, collapse='')) + coord_flip() + ggtitle(paste0("Host and parasite exonic regions overlaps")) + scale_fill_discrete(name = "", labels = c("Host overlaps", "Parasite overlaps")) + theme_light()
#   return(plot_gg)
#   }
# }
# 
# library(gridExtra)
# 
# pl <- lapply(studyIDs[c(1,2)], function(study) HostParasiteCountPyramid_ggplot(study))
# ml <- marrangeGrob(pl, nrow=length(studyIDs)/3, ncol=3)
# ml
```


<!-- 5. Regression of count variables: best as of now with glm, family poisson. Try negative binomial -->

<!-- ```{r} -->

<!-- ERP020067 <- allHPexpression[1:6,] -->
<!-- lm.trial <- glm(CountSum ~ InputReads + MapPercent + MultipleMapReads + UnmappedReads, data= ERP020067, family = "poisson") -->

<!-- summary(lm.trial) -->

<!-- ``` -->

\newpage
6. Parasite proportion vs # input reads for each hp pair.

```{r, echo = F, eval = T}
require(ggplot2)
hp_parasiteProportion_ggplot <- ggplot(allHPexpression, aes(log10(InputReads), ActualParasite, shape=HostParasite, color=Study))

hp_parasiteProportion_gg <- hp_parasiteProportion_ggplot + geom_point(alpha=0.7) + ylim(0,100) + ylab("Parasite proportion uniquely mapped") + xlab("Number of input reads in a sample (run)") + ggtitle("Parasite proportion relative to number of input reads for a host-parasite pair") + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18)) + theme_light() + scale_shape_manual(values = 0:7)
hp_parasiteProportion_gg
```

\newpage
7. Parasite proportion vs # input reads for each parasite and then for each host.

```{r, echo = F, eval = T}

hp_parasiteProportion_parasite_ggplot <- ggplot(allHPexpression, aes(log10(InputReads), ActualParasite, shape=Parasite, color=Parasite))

hp_parasiteProportion_parasite_gg <- hp_parasiteProportion_parasite_ggplot + geom_point(alpha=0.7) + ylim(0,100) + ylab("Parasite proportion uniquely mapped") + xlab("Number of input reads in a sample (run)") + ggtitle("Parasite proportion relative to number of input reads for each parasite") + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18)) + theme_minimal() + scale_shape_manual(values = 0:7)

hp_parasiteProportion_parasite_gg
```
\newpage
8. Make a summary table:


```{r, echo = F, eval = T}

Summary_table <- data.frame()

for(i in 1:length(studyIDs))
{
  total_runs <- nrow(read.table(paste0("/SAN/Plasmo_compare/SRAdb/Output/",studyIDs[i],"/runs_",studyIDs[i],".txt",collapse=''), header = F, sep = ','))
  table_study <- t(read.table(paste0("/SAN/Plasmo_compare/SRAdb/Output/",studyIDs[i],"/table_",studyIDs[i],".txt",collapse=''), header = T, sep = '\t'))

  Summary_table[i,1] <- studyIDs[i]
  Summary_table[i,2] <- total_runs
  Summary_table[i,3] <- (as.numeric(table_study[3])/total_runs)*100
  Summary_table[i,4] <- (as.numeric(table_study[4])/total_runs)*100
  Summary_table[i,5] <- (as.numeric(table_study[5])/total_runs)*100
  Summary_table[i,6] <- (as.numeric(table_study[6])/total_runs)*100
}

colnames(Summary_table) <- c("Study", "TotalRunsInStudy", "PercentageRunsIn70_30", "PercentageRunsIn90_10", "PercentageRunsIn99_1", "WorseThan99_1")

# for Latex:
# require(stargazer) # Hlavac, Marek (2018). stargazer: Well-Formatted Regression and Summary Statistics Tables.
# table_gg <- stargazer(Summary_table, rownames = FALSE, summary = FALSE, align = TRUE)

ggtexttable(Summary_table, rows = NULL, theme = ttheme("mBlue"))
# require(pander)
# table_gg <- knitr::kable(Summary_table)
# table_gg
write.table(Summary_table, "Summary_table.txt", sep='\t', row.names = F)
```
\newpage
Plot 9. Correlation between CountSum (total count given by countOverlaps() for a run == sum of genes in a run) and InputReads (total number of input reads for a run)

```{r, echo = F, eval = T}
cor(allHPexpression[,"InputReads"], allHPexpression[,"CountSum"])

CountSum_InputRead_cor_ggplot <- ggplot(allHPexpression, aes(x = log10(InputReads), y = log10(CountSum)))
CountSum_InputRead_cor_gg <- CountSum_InputRead_cor_ggplot + geom_point(alpha = 0.5) + geom_smooth() + theme_classic() + ggtitle("Correlation between overlapping exonic regions and number of input reads") + annotate("text", x = 7.8, y = 6.2, label = "R^2 == 0.80", parse = T, size = 8) + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), plot.title = element_text(size=18))
CountSum_InputRead_cor_gg
```

```{r, echo = F}
### suitability ###
suitability <- list()

number_of_df <- levels(as.factor(allExpression$HostParasite))

for(hp in number_of_df)
{
  # first get all runs for hp and remove rows with NA (for Pyoelii)
  hp.fulldf <- allExpression[allExpression$HostParasite==hp,]
  hp.fulldf <- hp.fulldf[,-c(13:14)]
  hp.fulldf <- na.omit(hp.fulldf)
  
  # select runs which have mapped above 65%
  hp.df <- hp.fulldf[as.numeric(hp.fulldf$MapPercent)>65,]
  
  # select runs with parasite reads greater than 10e6
  hp.df <- hp.df[as.numeric(hp.df$ParasiteSum)>1e6,]
  
  # select runs with host reads greater than 10e6
  hp.df <- hp.df[as.numeric(hp.df$HostSum)>1e6,]
  
  suitability[[hp]] <- hp.df
}

save(suitability, file = "suitability.RData")

Pfal<-allExpression[allExpression$Parasite=="Pfalciparum",]
Pber<-allExpression[allExpression$Parasite=="Pberghei",]
Pviv<-allExpression[allExpression$Parasite=="Pvivax",]
Pcha<-allExpression[allExpression$Parasite=="Pchabaudi",]
Pcyn<-allExpression[allExpression$Parasite=="Pcynomolgi",]
Pcoa<-allExpression[allExpression$Parasite=="Pcoatneyi",]
Pyoe<-allExpression[allExpression$Parasite=="Pyoelii",]
hPfal<-allExpression[allExpression$HostParasite=="humanPfalciparum",]
hPber<-allExpression[allExpression$HostParasite=="humanPberghei",]
hPviv<-allExpression[allExpression$HostParasite=="humanPvivax",]
mPcha<-allExpression[allExpression$HostParasite=="mousePchabaudi",]
mPyoe<-allExpression[allExpression$HostParasite=="mousePyoelii",]
mPber<-allExpression[allExpression$HostParasite=="mousePberghei",]
moPcyn<-allExpression[allExpression$HostParasite=="monkeyPcynomolgi",]
moPcoa<-allExpression[allExpression$HostParasite=="monkeyPcoatneyi",]

parasiteReads <- data.frame(Parasite = c("Pfalciparum", "Pberghei", "Pvivax", "Pchabaudi", "Pyoelii", "Pcynomolgi", "Pcoatneyi"), MappedReads10E9 = c(sum(Pfal$CountSum/10^9),sum(Pber$CountSum/10^9),sum(Pviv$CountSum/10^9),sum(Pcha$CountSum/10^9),sum(Pyoe$CountSum/10^9),sum(Pcyn$CountSum/10^9),sum(Pcoa$CountSum/10^9)))
hostparasiteReads <- data.frame(HostParasite = c("humanPfalciparum", "humanPberghei", "humanPvivax", "mousePchabaudi", "mousePberghei", "mousePyoelii", "monkeyPcynomolgi", "monkeyPcoatneyi"), MappedReads10E9 = c(sum(hPfal$CountSum/10^9),sum(hPber$CountSum/10^9),sum(hPviv$CountSum/10^9),sum(mPcha$CountSum/10^9),sum(mPber$CountSum/10^9),sum(mPyoe$CountSum/10^9),sum(moPcyn$CountSum/10^9),sum(moPcoa$CountSum/10^9)))
write.table(parasiteReads, "NumberOfReadsMappedOntoEachParasite.txt", sep = '\t', row.names = F)
write.table(hostparasiteReads, "NumberOfReadsMappedOntoEachHostParasite.txt", sep = '\t', row.names = F)

```

```{r}
require(ggplot2)

gg_h <- ggplot(allHPexp, aes(x = log10(ProteinCodHost), y = log10(NumberProtCodGenesHost)))
readsvsgenes <- gg_h + geom_point(alpha = 0.7, fill = allHPexp$colours) + xlab("log10 (Number of reads in a run mapping onto host genome)") + ylab("log10 (Number of genes represented by reads in a run)") + ggtitle("Number of reads mapping onto host genes")
readsvsgenes
```

```{r extend_allHPexp} 
require(dplyr)
# load allHPexp

human =
monkey =
mouse =

Pfalciparum =
Pvivax =
Pberghei = 
Pchabaudi =
Pyoelii =
Pcynomolgi =
Pcoatneyi =

Org.db = data.frame(human = human, monkey = monkey, mouse = mouse, Pfalciparum = Pfalciparum,
	Pvivax = Pvivax, Pberghei = Pberghei, Pchabaudi = Pchabaudi, Pyoelii = Pyoelii,
	Pcynomolgi = Pcynomolgi, Pcoatneyi = Pcoatneyi)

TotalHostGenes = c()
TotalParaGenes = c()

for(i in 1:nrow(allHPexp))
{
	TotalParaGenes[i] <- Org.db[grep(pattern = allHPexp[i, "Parasite"], colnames(Org.db))]
	TotalHostGenes[i] <- Org.db[grep(pattern = allHPexp[i, "Host"], colnames(Org.db))]
}

allHPexp[ncol(allHPexp)+1] <- TotalHostGenes
allHPexp[ncol(allHPexp)+1] <- TotalParaGenes

# include orthogroup related counts

# use positive experiments and _coding_genes.txt to create ortho version of each study
positive_experiments <- 
studyIDs <- 
parasite_orthogroups <-
host_orthogroups <-

for(i in 1:length(studyIDs))
{
	#if the study_coding_genes exists, merge with orthogroups (join functions)
	filepath = paste0("/SAN/Plasmo_compare/SRAdb/Output/",studyIDs[i],"/",studyIDs[i],"_coding_genes.txt", collapse = "")
	if(file.exists(filepath))
	{
		# find out what host and parasite the study is
		host <- positive_experiments[grep(pattern = studyIDs[i], positive_experiments[,1]),2]
		para <- positive_experiments[grep(pattern = studyIDs[i], positive_experiments[,1]),3]

		# take the host and para orthogroups and make a df -> orthogroup | gene name 

		h.df <- data.frame(Orthogroup = host_orthogroups[,1], Org = host_orthogroups[,grep(pattern = host, colnames(host_orthogroups))])
		p.df <- data.frame(Orthogroup = para_orthogroups[,1], Org = para_orthogroups[,grep(pattern = para, colnames(para_orthogroups))])

		hp.df <- cbind(h.df, p.df)

		# read table
		file = read.delim(filepath, header = T) %>% tibble::rownames_to_column("Gene")

		ortho.table = merge(file, hp.df, by.x = "Gene", by.y = "Org")

		write.table(ortho.table, paste0("/SAN/Plasmo_compare/SRAdb/Output/",studyIDs[i],"/",studyIDs[i],"_orthogroups.txt", collapse = ""))
	}

}

# add new study orthogroups to ortho_data.txt
# merge study_orthogroups.txt to ortho_data.txt

ortho_data <- read.delim("/SAN/Plasmo_compare/SRAdb/Output/ortho_data.txt",header = T)
ortho_data <- merge(ortho_data, SRP250329_orthogroups.txt)

 # % of orthogroups expressed
OrthoHostGenesExp = c()
OrthoParaGenesExp = c()

for(j in 1:nrow(allHPexp))
{
	study <- allHPexp[j, "Study"]
	
	# read file of study_orthogroups file
	countfile <- read.delim(paste0("/SAN/Plasmo_compare/SRAdb/Output/",study,"/", study,"_orthogroups.txt", collapse = ""), header = T) %>% 
		tibble::column_to_rownames("Orthogroup") %>%
		select(-c("Gene", "Org"))

	study_count <- c()

	for(x in 1:ncol(countfile))
	{
		study_count[x] <- length(which(countfile[,x] > 0))
		names(study_count[x]) <- colnames(countfile)[x]

	}
}
```
